### PR TITLE
[SYCL] Add logics for aligned_alloc_xxx<T> to deal with unsupported Alignment argument

### PR DESCRIPTION
**Pull request by HPS-1**
Added logics for aligned_alloc_xxx\<T\> to return nullptr when the Alignment argument is not a power of 2. Also updated and added a new test file align.cpp used to test such cases (where Alignment is not 2's power). The fixed methods, all in sycl/include/sycl/usm.hpp, are:
aligned_alloc_device\<T\>
aligned_alloc_shared\<T\>
aligned_alloc_host\<T\>
aligned_alloc\<T\>

*Note that each of these methods has two slighly different versions with different parameter lists.

**Also note that Alignment=0 is considered as a valid "default" case in our implementation so it will not cause nullptr to be returned. (Code here: https://github.com/intel/llvm/blob/435845bc23f68ad146915018aa5dd5b451d8bc5e/sycl/source/detail/usm/usm_impl.cpp#L163) (Also if we let these methods return nullptr when Alignment=0, test SYCL :: Annotated_usm/annotated_usm_kind.cpp will fail.)

Link to the issue: https://github.com/intel/llvm/issues/11642

### CODE REVIEW

File: sycl/include/sycl/usm.hpp [hunk 1]
```diff
@@ -179,18 +179,26 @@ T *aligned_alloc_device(
     size_t Alignment, size_t Count, const device &Dev, const context &Ctxt,
     const property_list &PropList = {},
     const detail::code_location &CodeLoc = detail::code_location::current()) {
-  return static_cast<T *>(aligned_alloc_device(max(Alignment, alignof(T)),
-                                               Count * sizeof(T), Dev, Ctxt,
-                                               PropList, CodeLoc));
+  if (!(Alignment & (Alignment - 1))) {
```
[the remark this answers is not in the document]
Comment (HPS-1) [reply sycl/include/sycl/usm.hpp:182]: We agree that the current behavior should be leaved unchanged. We may open a new issue around this if necessary.

File: sycl/test-e2e/USM/align.cpp [hunk 2]
```diff
@@ -0,0 +1,108 @@
+// RUN: %{build} -o %t.out
+// RUN: %{run} %t.out
+
+// UNSUPPORTED: gpu
+
+// E2E tests for annotated USM allocation functions with alignment arguments
+// that are not powers of 2. Note this test does not work on gpu because some
+// tests expect to return nullptr, e.g. when the alignment argument is not a
+// power of 2, while the gpu runtime has different behavior
```
[the remark this answers is not in the document]
Comment (HPS-1) [reply sycl/test-e2e/USM/align.cpp]: This is a piece of outdated comment and has been removed in the latest revision. I believe the original author misunderstood the reason of the problem they encountered - it's caused the bugs (i.e. lacking logics to deal with Alignment arguments that are not 2's powers) I fixed in this PR, not something about GPU runtime.

**Review by bso-intel — APPROVED** [review]

File: sycl/test-e2e/USM/align.cpp [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+      [&]() { return ADevice(3, q); }, [&]() { return ADevice(5, dev, Ctx); },
+      [&]() { return AHost(7, q); }, [&]() { return AHost(9, Ctx); },
+      [&]() { return AShared(114, q); },
+      [&]() { return AShared(1023, dev, Ctx); },
+      [&]() { return AAnnotated(15, q, alloc::device); },
+      [&]() { return AAnnotated(17, dev, Ctx, alloc::host); }
+      // Case: aligned_alloc_xxx<T> with no alignment property, and the
+      // alignment
+      // argument is not a power of 2, the result is nullptr
+      ,
```
Comment (AlexeySachkov) [inline sycl/test-e2e/USM/align.cpp:87]:
```suggestion
      [&]() { return AAnnotated(17, dev, Ctx, alloc::host); },
      // Case: aligned_alloc_xxx<T> with no alignment property, and the
      // alignment argument is not a power of 2, the result is nullptr
```

File: sycl/test-e2e/USM/align.cpp [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+  };
+  auto ATAnnotated = [&](size_t align, auto... args) {
+    return aligned_alloc<T>(align, N, args...);
+  };
+
+  // Test cases that are expected to return null
+  auto check_null = [&q](auto AllocFn, int Line, int Case) {
+    decltype(AllocFn()) Ptr = AllocFn();
+    auto v = reinterpret_cast<uintptr_t>(Ptr);
+    if (v != 0) {
```
Comment (AlexeySachkov) [inline sycl/test-e2e/USM/align.cpp:58]: Why don't we compare it with `nullptr` directly?

File: sycl/test-e2e/USM/align.cpp [hunk 5]
```diff
@@ -0,0 +1,102 @@
+// RUN: %{build} -o %t.out
+// RUN: %{run} %t.out
+
+// UNSUPPORTED: gpu
+
+// E2E tests for annotated USM allocation functions with alignment arguments
```
Comment (AlexeySachkov) [inline sycl/test-e2e/USM/align.cpp:6]: Unified Runtime [documentation says](https://oneapi-src.github.io/unified-runtime/core/api.html#urusmdevicealloc) that there should be an error in that case. I think that we should submit a bug report to the corresponding GitHub repo if we don't see that error.

File: sycl/test-e2e/USM/align.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (HPS-1) [reply sycl/test-e2e/USM/align.cpp]: Respective issue opened here: https://github.com/intel/llvm/issues/12638

File: sycl/test-e2e/USM/align.cpp (hunk 5 above)
Comment (HPS-1) [reply sycl/test-e2e/USM/align.cpp:6]: Respective issue opened here: https://github.com/intel/llvm/issues/12638

**Pull request merged**